### PR TITLE
Fix multiprocessing cache (content changed, but ptr is still valid)

### LINKED ISSUE #16141: [multiprocessing] does not play well with distributions in GPU
## 🐛 Bug

Continued from some discussions at https://github.com/pytorch/pytorch/pull/14736 and https://github.com/uber/pyro/pull/1694.

Running the following script gives the error **RuntimeError: Assertion `self->allocator() != nullptr' failed**:
```sh
Traceback (most recent call last):
  File "testt.py", line 31, in <module>
    sample = q.get()
  File "/home/fehiepsi/miniconda3/envs/pydata/lib/python3.6/multiprocessing/queues.py", line 113, in get
    return _ForkingPickler.loads(res)
  File "/home/fehiepsi/miniconda3/envs/pydata/lib/python3.6/site-packages/torch/multiprocessing/reductions.py", line 104, in rebuild_cuda_tensor
    t = torch._utils._rebuild_tensor(storage, tensor_offset, tensor_size, tensor_stride)
  File "/home/fehiepsi/miniconda3/envs/pydata/lib/python3.6/site-packages/torch/_utils.py", line 131, in _rebuild_tensor
    return tensor_class().set_(storage, storage_offset, size, stride)
RuntimeError: Assertion `self->allocator() != nullptr' failed.  at /opt/conda/conda-bld/pytorch-nightly_1547631207535/work/aten/src/THC/THCStorage.cpp:16
```

## To Reproduce

Running the following script

```
import torch
import torch.distributions as dist
import torch.multiprocessing as mp

torch.set_default_tensor_type(torch.cuda.FloatTensor)
n = 10

def model():
    d1 = dist.Normal(torch.zeros(3), 1)
    v1 = d1.rsample()  # no error when replace `.rsample()` by `.sample()`
    d2 = dist.Normal(v1, 2)
    v2 = d2.rsample()
    return [(d1, v1), (d2, v2)]

def worker(q, e):
    for i in range(n):
        sample = [torch.zeros(1), torch.ones(1)]  # no error when remove this line ???
        sample = model()
        q.put(sample)
        e.wait()
        e.clear()

if __name__ == "__main__":
    ctx = mp.get_context("spawn")
    q = ctx.Queue()
    e = ctx.Event()
    p = ctx.Process(target=worker, args=(q, e))
    p.start()
    for i in range(n):
        print("=== ITER {} ===".format(i))
        sample = q.get()
        print(sample)
        e.set()
    p.join()
```

## Expected behavior

No error as when running in CPU.

## Environment

 - PyTorch Version (e.g., 1.0): 1.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: 9
 - GPU models and configuration: GTX 1080

## Additional context

The error does not happen with PyTorch 1.0.dev.20181202 version.

cc @neerajprad @ailzhang @ezyang

### PR DESCRIPTION
**Pull request by ailzhang**
Fix #16141

### CODE REVIEW

**Comment by ezyang** [comment]
Yeah, this is super hacky. Here is a simpler approach that will give better guarantees: what if we associate every storage with a unique monotonically increasing number (storing it in the storage) and use this for indexing? The very simplest implementation would just be an atomic counter but there are also fancier schemes that can avoid syncing across cpu.

**Comment by ezyang** [comment]
I think that cannot happen, because storage handle should be different in those cases.

**Comment by ailzhang** [comment]
@ezyang  I added the `epoch` implementation but there's one thing I'm still a little confused. In a simple example script we had for this issue, our `StorageImpl` constructor was called over 1000 times while most of them were constructing `numel=0` storages. I'm not sure if this is expected behavior or something is wrong in my implementation. 
I left some comments in the commit. You can see the example in `python test_multiprocessing.py TestMultiprocessing.test_cuda_storage_sharing_epoch > log`
Running the same script on `master` also has a similar behavior. With these many calls, I'm a bit worried about overflowing `int32_t` and performance degrading due to atomic ops. Do you know the cause of this? Thanks!

**Comment by ezyang** [comment]
> most of them were constructing numel=0 storages

I don't find this too surprising: a lot of operators operate by first allocating a zero-size tensor, and then resizing it later.

As for whether or not we expect the constructor to be called a 1000 times, I am not too sure. But it would not *surprise* me per se that they are allocating that many tensors.

On the subject of overflow: the intent of this change is not to be 100% correct; it's just to make the failure case less likely to occur. We probably *should* wrap around when we run out of space; this suggests `uint32_t` is more appropriate (since overflow is only defined on unsigned integers.)

File: c10/core/StorageImpl.cpp [hunk 1]
```diff
@@ -1 +1,28 @@
 #include <c10/core/StorageImpl.h>
+
+namespace c10 {
+  std::atomic<int> global_counter (0);
```
Comment (ezyang) [inline c10/core/StorageImpl.cpp]: We should put this in an anonymous namespace or something like that.

File: c10/core/StorageImpl.cpp [hunk 2]
```diff
@@ -1 +1,28 @@
...
+        allocator_(allocator) {
+    if (numel > 0) {
+      if (data_type_.id() == caffe2::TypeIdentifier::uninitialized()) {
+        AT_ERROR(
+            "Constructing a storage with meta of unknown type and non-zero numel");
+      }
+    }
+    epoch_ = global_counter.load();
+    std::cout << "creating storage " << epoch_ << " with size " << numel_ << std::endl;
+    global_counter++;
```
Comment (ezyang) [inline c10/core/StorageImpl.cpp]: Use `fetch_add` to do the read and write atomically.

**Comment by ezyang** [comment]
Let's talk about how to avoid performance degradation from atomic op.

First, we shouldn't optimize unless we have some numbers to back ourselves up. The situation we expect to be slowed down by the addition of an atomic counter is when we have many threads allocating many storages in rapid succession. This will cause bad cache behavior because the cache line for the atomic counter will keep bouncing between threads. So if this is actually a problem, we expect a synthetic benchmark like this to get slower when we add an atomic counter here.

Second, supposing there is a slow down, how can we fix it? Observe that we don't actually need all of the guarantees that an atomic counter would give us: all we need is a unique *enough* identifier for each storage. (And it's OK if we have two allocations in close succession hit the same identifier, because they are unlikely to alias the same memory in that case.) So we could, for example, use RDTSC to read out a cycle count that we'll associate with the storage. This particular instruction is super cheap (but it's not available on all CPUs, so you'll still need some fallback code--thus the importance of checking!)

cc @colesbury  if you have any opinions

**Comment by colesbury** [comment]
I'm not sure this is right... the bug in the original code is that it the imported tensor is alive when the original tensor is deleted.  (See https://pytorch.org/docs/stable/multiprocessing.html#sharing-cuda-tensors)

If you change:

```python
sample = q.get()
print(sample)
e.set()
```

to

```python
sample = q.get()
print(sample)
del sample
e.set()
```

You'll notice that it doesn't fail

**Comment by ailzhang** [comment]
Thanks @colesbury for pointing out the error in the script! This is indeed due to the alive `sample` on the receiver side. I mistakenly thought `sample` should go out of scope using a C intuitive... @ezyang sorry for the confusion.
I will go ahead and close this PR and edit my reply in the original issue. 

In general it should be helpful to add a code example in https://pytorch.org/docs/stable/multiprocessing.html#sharing-cuda-tensors to list all current constraints of our cuda multiprocessing implementation, and provide a guideline on how to use it properly. I will check to update the doc for this part.

**Pull request closed without merging**